operator/v1alpha1/types_externaldns.go

```diff
@@ -0,0 +1,397 @@
+
+// ExternalDNSUnion describes optional fields for an ExternalDNS Source that should
+// be captured.
+// +union
+type ExternalDNSSourceUnion struct {
+	// Type specifies an ExternalDNS source resource
+	// to create DNS records for.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	Type ExternalDNSSourceType `json:"type"`
+
+	// AnnotationFilter describes an annotation filter
+	// used to filter which source instance resources
+	// ExternalDNS publishes records for.
+	// The annotation filter uses label selector semantics
+	// against source resource annotations.
+	//
+	// +optional
+	AnnotationFilter map[string]string `json:"annotationFilter,omitempty"`
+
+	// Namespace instructs ExternalDNS to only acknowledge
+	// source resource instances in a specific namespace.
+	//
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Service describes source configuration options specific
+	// to the service source resource.
+	//
+	// +optional
+	Service *ExternalDNSServiceSourceOptions `json:"service,omitempty"`
+
+	// CRD describes source configuration options specific
+	// to the CRD source resource.
+	//
+	// +optional
+	CRD *ExternalDNSCRDSourceOptions `json:"crd,omitempty"`
+
+	// OpenShiftRoute source currently has no unique configuration options.
+}
+
+// +kubebuilder:validation:Enum=OpenShiftRoute;Service;CRD
+type ExternalDNSSourceType string
+
+const (
+	SourceTypeRoute   ExternalDNSSourceType = "OpenShiftRoute"
+	SourceTypeService ExternalDNSSourceType = "Service"
+	SourceTypeCRD     ExternalDNSSourceType = "CRD"
+)
+
+// +kubebuilder:validation:Enum=Ignore;Allow
+type HostnameAnnotationPolicy string
+
+const (
+	HostnameAnnotationPolicyIgnore HostnameAnnotationPolicy = "Ignore"
+	HostnameAnnotationPolicyAllow  HostnameAnnotationPolicy = "Allow"
+)
+
+// ExternalDNSServiceSourceOptions describes options
+// specific to the ExternalDNS service source.
+type ExternalDNSServiceSourceOptions struct {
+	// PublishPolicy determines what kinds of Service
+	// resources are published to the given DNS provider.
+	//
+	// The following publishing policies are available:
+	//
+	//  "PublishExternal": Only publish DNS records for
+	//  service that are externally reachable,
+	//  such as NodePort, ExternalName, and LoadBalancer
+	//  services. This is the default behavior of ExternalDNS.
+	//
+	//  "PublishInternalExternal": Publish DNS records
+	//   for externally reachable services, in addition to
+	//   internally reachable services, such as ClusterIP
+	//   services.
+	//
+	// +optional
+	PublishPolicy *ServicePublishPolicy `json:"publishPolicy,omitempty"`
+
+	// ServiceType determines what types of Service resources
+	// are watched by ExternalDNS. The following types are
+	// available options:
+	//
+	//  "NodePort"
+	//  "ExternalName"
+	//  "LoadBalancer"
+	//  "ClusterIP" (requires a publishPolicy of "PublishInternalExternal")
+	//
+	// One or more Service types can be specified, if desired.
```

Comment:
Document what the behavior is if this list is empty.  Alternatively, add `// +kubebuilder:validation:MinLength=1` if an empty list is invalid.

Reply:
I went ahead and made this field required with a kubebuilder default of "LoadBalancer".
WDYT?

Reply:
That seems like a reasonable default.  Does that make `// +kubebuilder:validation:MinLength=1` superfluous, or could the user still specify an explicit empty value that defaulting wouldn't override?